### PR TITLE
When rehashing a Set or Map, reduce the peak memory used.

### PR DESCRIPTION
**Pull request by erikcorry**
For large sets or maps we build a new index from scratch when we outgrow the old one.  In this case we reuse the arraylets from the old index, rather than creating new arraylets and keeping the old ones alive during rehashing.

This also changes the rules so that we go to LargeArray_ as soon as we exceed the arraylet size.  Previously we stayed at SmallArray_ until we hit the max size for page allocations.  On embedded this will avoid arrays in the 2-3.9k size range, which could waste memory on fragmentation.

### CODE REVIEW

File: lib/core/collections.toit [hunk 1]
```diff
@@ -845,12 +845,8 @@ abstract class Array_ extends List:
   // Optimized helper method for iterating over the array elements.
   abstract do_ end/int [block] -> none
 
-  /** Create a new array, copying up to old_length elements from this array. */
-  resize_for_list_ old_length/int new_length/int -> Array_:
-    result := Array_ new_length
-    for i := 0; i < old_length and i < new_length; i++:
-      result[i] = this[i]
-    return result
+  /** Create a new array, copying up to $copy_length elements from this array. */
```
Comment (kasperl) [inline lib/core/collections.toit:848]: Use /// for consistency?

Also why do we call this copy_length and new_length here and not copy_size and new_size?

File: lib/core/collections.toit [hunk 2]
```diff
@@ -970,24 +970,31 @@ class LargeArray_ extends Array_:
...
-        if it * ARRAYLET_SIZE < old_size:
+        arraylet = Array_ limit filler
+        if pos < copy_size:
           old_arraylet := vector_[it]
-          arraylet.replace 0 old_arraylet 0 (old_size % ARRAYLET_SIZE)
-        left -= limit
+          arraylet.replace 0 old_arraylet 0 (min limit (copy_size - pos))
+      new_vector[it] = arraylet
+      left -= limit
+      pos += ARRAYLET_SIZE
```
Comment (kasperl) [inline lib/core/collections.toit:997]: It feels odd to me that you're increasing/decreasing left and pos with what appears to be different amounts. Why not use limit consistently?

File: lib/core/collections.toit [hunk 3]
```diff
@@ -970,24 +970,31 @@ class LargeArray_ extends Array_:
...
+  // An expansion or shrinking that uses the backing arraylets from the old
+  // array.  Used by List, Set, and Map.
+  resize_for_list_ copy_size/int new_size/int filler:
     // Rounding up division.
     number_of_arraylets := ((new_size - 1) / ARRAYLET_SIZE) + 1
-    new_vector := Array_ number_of_arraylets
+    if number_of_arraylets == 1:
+      return vector_[0].resize_for_list_ copy_size new_size filler
+    new_vector := Array_ number_of_arraylets  // new_vector may be a LargeArray_!
     left := new_size
```
Comment (kasperl) [inline lib/core/collections.toit:981]: remaining?

File: lib/core/collections.toit [hunk 4]
```diff
@@ -970,24 +970,31 @@ class LargeArray_ extends Array_:
...
+        arraylet = vector_[it]
+        if pos + limit > copy_size:
+          arraylet.fill --from=(max 0 (copy_size - pos)) --to=limit filler
       else:
-        limit := min left ARRAYLET_SIZE
-        arraylet := Array_ limit
-        new_vector[it] = arraylet
-        if it * ARRAYLET_SIZE < old_size:
+        arraylet = Array_ limit filler
+        if pos < copy_size:
```
Comment (kasperl) [inline lib/core/collections.toit:992]: Is this check necessary and worth it?

File: lib/core/collections.toit [hunk 5]
```diff
@@ -970,24 +970,31 @@ class LargeArray_ extends Array_:
...
+    pos := 0
     number_of_arraylets.repeat:
-      if (it + 1) * ARRAYLET_SIZE <= old_size:
-        new_vector[it] = vector_[it]
-        left -= ARRAYLET_SIZE
+      arraylet := ?
+      limit := min left ARRAYLET_SIZE
+      if pos + limit <= size:
+        arraylet = vector_[it]
+        if pos + limit > copy_size:
```
Comment (kasperl) [inline lib/core/collections.toit:988]: Is this check necessary and worth it?

File: lib/core/collections.toit [hunk 6]
```diff
@@ -2168,10 +2176,19 @@ abstract class HashedInsertionOrderedCollection_:
...
+      // Rebuild the index using the backing array.
+      // By using resize_for_list_ we reuse the arraylets when growing large
+      // arrays.  This reduces GC churn and, more importantly, peak memory
+      // usage.
+      if index_:
+        index_ = index_.resize_for_list_ /*copy_size=*/ 0 new_index_size /*filler=*/0
+        index_.fill 0
+      else:
+        index_ = Array_ new_index_size 0
+      if index_.size != new_index_size: throw "foo"
```
Comment (kasperl) [inline lib/core/collections.toit:2191]: Should this be an assert?

File: src/primitive_core.cc [hunk 7]
```diff
@@ -1601,17 +1601,16 @@ PRIMITIVE(array_at_put) {
...
 // the new one.
 PRIMITIVE(array_expand) {
-  ARGS(Array, old, word, old_length, word, length);
+  ARGS(Array, old, word, old_length, word, length, Object, filler);
   if (length == 0) return process->program()->empty_array();
   if (length < 0) OUT_OF_BOUNDS;
-  if (length > Array::max_length_in_process()) OUT_OF_RANGE;
-  if (old_length < 0 || old_length > old->length() || old_length > length) OUT_OF_RANGE;
-  Object* result = process->object_heap()->allocate_array(length, Smi::from(0));
+  if (length > Array::ARRAYLET_SIZE) OUT_OF_RANGE;
```
Comment (kasperl) [inline src/primitive_core.cc:1607]: ARRAYLET_SIZE is 500 right now. Is that a good number? Can we make it much larger on 64-bit platforms where the page size is bigger?

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: lib/core/collections.toit (hunk 4 above)
Comment (erikcorry) [reply lib/core/collections.toit:992]: replace doesn't allow negative values for the `to` argument, so yes.

File: lib/core/collections.toit (hunk 5 above)
Comment (erikcorry) [reply lib/core/collections.toit:988]: It's hard to measure a difference.  I took it out.

**Pull request merged**